contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,259 @@
+        _carbonController = carbonController;
+        _bancorNetwork = bancorNetwork;
+        _bnt = bnt;
+    }
+
+    /**
+     * @dev fully initializes the contract and its parents
+     */
+    function initialize() external initializer {
+        __CarbonVortex_init();
+    }
+
+    // solhint-disable func-name-mixedcase
+
+    /**
+     * @dev initializes the contract and its parents
+     */
+    function __CarbonVortex_init() internal onlyInitializing {
+        __Upgradeable_init();
+        __ReentrancyGuard_init();
+
+        __CarbonVortex_init_unchained();
+    }
+
+    /**
+     * @dev performs contract-specific initialization
+     */
+    function __CarbonVortex_init_unchained() internal onlyInitializing {
+        setRewardsPPM(100_000);
+    }
+
+    /**
+     * @dev authorize the contract to receive the native token
+     */
+    receive() external payable {}
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure override(IVersioned, Upgradeable) returns (uint16) {
+        return 2;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function setRewardsPPM(uint256 newRewardsPPM) public onlyAdmin validFee(uint32(newRewardsPPM)) {
+        uint256 prevRewardsPPM = _rewardsPPM;
+
+        // return if the rewards percentage PPM is the same
+        if (prevRewardsPPM == newRewardsPPM) {
+            return;
+        }
+
+        _rewardsPPM = newRewardsPPM;
+
+        emit RewardsUpdated({ prevRewardsPPM: prevRewardsPPM, newRewardsPPM: newRewardsPPM });
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function rewardsPPM() external view returns (uint256) {
+        return _rewardsPPM;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function totalBurned() external view returns (uint256) {
+        return _totalBurned;
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function availableFees(Token token) external view returns (uint256) {
+        return _carbonController.accumulatedFees(token) + token.balanceOf(address(this));
+    }
+
+    /**
+     * @inheritdoc ICarbonVortex
+     */
+    function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+        // cache rewardsPPM to save gas
+        uint256 rewardsPercentage = _rewardsPPM;
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
```

Comment:
`withdrawFees` returns the withdrawn amount, so perhaps it can be used to avoid calling `balanceOf`. It won't handle the case of "accidentally stuck" tokens, but:
1. It'll decrease gas costs nicely.
2. These can be handled by a dust collecting function that can be added to the vortex in the future.

Reply:
Actually the reason we use balanceOf is that the vortex will be used to store the rewards coming from the fast lane. They will be burned through the vortex.